### PR TITLE
feat(ecs): add codepipeline deploy-action to ecs cluster

### LINKED ISSUE #1386: Add ECS Deploy Action for Codepipeline
Codepipeline can deploy to ESC-Cluster. Package `aws-ecs` need connectors for codepipeline, like `pipeline-actions.ts` in `aws-ecr`, but extends from `codepipeline-api.DeployAction`.

Allowed Input Artifacts: 1 [(image definition file)](https://docs.aws.amazon.com/codepipeline/latest/userguide/pipelines-create.html#pipelines-create-image-definitions)
Allowed Output Artifacts: 0

See [(Action Type Constraints for Artifacts)](https://docs.aws.amazon.com/codepipeline/latest/userguide/reference-pipeline-structure.html#action-requirements)  for all codepipeline actions

### PR DESCRIPTION
**Pull request by aweiher**
No Unit-Test, one Integration Test ist present. The similar compontent `s3 deploy-action` has no unit-test (didnt found it), so I oriented there. Please give me feedback when required.

I have tested the integration test in my private AWS Account, and it worked.

Notes:
- I did not implement `toCodePipelineInvokeAction` (on service?) as `service` and `repository` is required, as I didnt figured out an easy api for this. Need feedback if this is required.
- The default IAM rules are maybe too open (`ecs:*` for `*`), I oriented on the default AWS CodePipeline Role: https://docs.aws.amazon.com/codepipeline/latest/userguide/how-to-custom-role.html
- Not included in the component, but in the integration-test: `sts:AssumeRole` and `iam:PassRole` for `*` - same reason as point before. The following would be ideal, but I didnt figured out how to construct this with cdk: 

```yaml
              - Resource: "*"
                Effect: Allow
                Action:
                  - iam:PassRole
                Condition:
                  StringEqualsIfExists:
                    iam:PassedToService:
                      - ec2.amazonaws.com
                      - ecs-tasks.amazonaws.com
```

Fixes #1386

----

### Pull Request Checklist

* [ ] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

**Comment by rix0rrr** [comment]
@skinny85, @SoManyHs... opinions?

File: packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-ecs-deploy.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    new ecs.EcsDeployAction({
+        actionName: 'DeployAction',
+        inputArtifact: sourceAction.outputArtifact,
+        fileName: 'imageDefinition.json',
+        service,
+        repository
+    })
+);
+
+pipeline.addToRolePolicy(
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-ecs-deploy.ts]: Why is this needed?

File: packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-ecs-deploy.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
+import ec2 = require('@aws-cdk/aws-ec2');
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-ecs-deploy.ts]: The indentation in this file should be 2 spaces (I see it is in the other files, so this is probably a simple settings snafu).

File: packages/@aws-cdk/aws-ecs/lib/base/pipeline-action.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+import { PolicyStatement } from '@aws-cdk/aws-iam';
+import cdk = require('@aws-cdk/cdk');
+import { BaseService } from './base-service';
+
+/**
+ * Common properties for the {@link PipelineSourceAction CodePipeline deploy Action},
+ * whether creating it directly,
+ * or through the {@link IRepository#toCodePipelineSourceAction} method.
+ */
+export interface PipelineDeployActionProps extends codepipeline.CommonActionProps {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-ecs/lib/base/pipeline-action.ts]: Were you planning on adding a `toCodePipelineDeployAction` method on the Service class? Because if not, the documentation on this interface seems incorrect :) (it would be cool to have that method on Service though 🙂).

File: packages/@aws-cdk/aws-ecs/lib/base/pipeline-action.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+  /**
+   * the service to deploy inside the ECS cluster
+   */
+  service: BaseService
+
+  /**
+   * the repository where the docker image(s) are stored
+   */
+  repository: Repository
```
Comment (skinny85) [inline packages/@aws-cdk/aws-ecs/lib/base/pipeline-action.ts]: I'm wondering: should this property be optional? It seems like it's only used for granting permissions.

Alternatively, if there's a way to get this object from the taskDefinition contained in the Service, that would be idea.

File: packages/@aws-cdk/aws-ecs/lib/base/pipeline-action.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+ */
+export interface PipelineDeployActionProps extends codepipeline.CommonActionProps {
+  /**
+   * the input artifact contains the image definitions file which describes the ECR docker image(s) to deploy
+   * @see https://docs.aws.amazon.com/codepipeline/latest/userguide/pipelines-create.html#pipelines-create-image-definitions
+   */
+  inputArtifact: codepipeline.Artifact;
+
+  /**
+   * the filename of the image definitions file in the input artifact
```
Comment (skinny85) [inline packages/@aws-cdk/aws-ecs/lib/base/pipeline-action.ts]: Can you use full sentences for the JSDocs (begins with an uppercase letter, ends with a period)? Thanks!

File: packages/@aws-cdk/aws-ecs/lib/base/pipeline-action.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  /**
+   * the repository where the docker image(s) are stored
+   */
+  repository: Repository
+}
+
+/**
+ * CodePipelie Deploy Action to deploy the Service to ECS Cluster
+ */
+export class EcsDeployAction extends codepipeline.DeployAction {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-ecs/lib/base/pipeline-action.ts]: According to our current conventions, the name of this class should be `PipelineDeployAction`, so that it reads `ecs.PipelineDeployAction` when used.

File: packages/@aws-cdk/aws-ecs/lib/base/pipeline-action.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+    stage.pipeline.role
+      .addToPolicy(
+        new PolicyStatement()
+          .addAction(
+            "ecs:*"
+          )
+          .addAllResources()
+      );
+
```
Comment (skinny85) [inline packages/@aws-cdk/aws-ecs/lib/base/pipeline-action.ts]: This empty line can be removed.

File: packages/@aws-cdk/aws-ecs/lib/base/pipeline-action.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    // see: https://docs.aws.amazon.com/codepipeline/latest/userguide/how-to-custom-role.html
+    this.repository.grantPull(stage.pipeline.role);
+
+    stage.pipeline.role
+      .addToPolicy(
+        new PolicyStatement()
+          .addAction(
+            "ecs:*"
+          )
+          .addAllResources()
```
Comment (skinny85) [inline packages/@aws-cdk/aws-ecs/lib/base/pipeline-action.ts]: Is there any way to scope the permissions down somewhat?

File: packages/@aws-cdk/aws-ecs/README.md [hunk 9]
```diff
@@ -291,6 +291,26 @@ To start an Amazon ECS task on an Amazon EC2-backed Cluster, instantiate an
 
 > Note: it is currently not possible to start AWS Fargate tasks in this way.
 
+## Integration with CodePipeline
+
+CodePipeline can deploy to the ECS Cluster. The deploy action stage receives 
+one input-artifact which contains the [image definition file].
+
+```ts
+const deployStage = pipeline.addStage({ name: "Deploy" });
```
Comment (skinny85) [inline packages/@aws-cdk/aws-ecs/README.md]: You can do it in one expression:

```ts
const deployStage = pipeline.addStage({
  name: "Deploy",
  actions: [
    new ecs.EcsDeployAction({
      actionName: 'DeployAction',
      inputArtifact: sourceAction.outputArtifact,
      fileName: 'imageDefinition.json',
      service,
      repository
    }),
  ],
});
```

**Comment by skinny85** [comment]
The build failure seems like something transient. Can you rebase, and try again? (there are some conflicts with current master, so that would have to be done sometime anyway). Thanks!

**Comment by aweiher** [comment]
Hej @skinny85 - Thank you very much for your code-review! :) I will fix the issues in the next days :+1:

**Comment by skinny85** [comment]
Hey @aweiher ,

there's been a pretty big revolution in the way the Actions are represented in code in PR #2098 . This might make rebasing this PR particularly challenging. Given how close to the final shape this PR is, I'd be willing to take it over from you - I was the one doing the changes in #2098, so I'm quite familiar with what needs to be changed in this new way of doing things.

Let me know!

File: packages/@aws-cdk/aws-ecs/lib/base/pipeline-action.ts (hunk 8 above)
Comment (aweiher) [reply packages/@aws-cdk/aws-ecs/lib/base/pipeline-action.ts]: Just added minimal permissions for  ECR source action.

**Comment by aweiher** [comment]
Hi @skinny85 - No Problem, take this PR over. I really wanted to implement this by myself, but an other task is eating all my time unfortunately. I added one commit/comment regarding the permission, hope it helps.

File: packages/@aws-cdk/aws-ecs/lib/base/pipeline-action.ts (hunk 4 above)
Comment (aweiher) [reply packages/@aws-cdk/aws-ecs/lib/base/pipeline-action.ts]: Afaik it is not possible to get the repository from the service (its possible to update the service from different repositories)

File: packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-ecs-deploy.ts (hunk 1 above)
Comment (aweiher) [reply packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-ecs-deploy.ts]: This is required for the ecr source-action to watch for changes in the repository. 
I moved this in the last commit to the `pipeline-action`, think its the better place for this line. 

As written above, in the original IAM statement, the following cloudformation construct is included:

```yaml
            - Resource: "*"
              Effect: Allow
              Action:
                - iam:PassRole
              Condition:
                StringEqualsIfExists:
                  iam:PassedToService:
                    - ec2.amazonaws.com
                    - ecs-tasks.amazonaws.com
```

But I didn't figured out how to construct this with cdk.

**Comment by skinny85** [comment]
Thanks, will do!

**Comment by skinny85** [comment]
I've rebased the PR on top of master. @RomainMuller and @SoManyHs , I would appreciate if you took a look.

**Comment by skinny85** [comment]
Missed a new feature in ECS when rebasing that changes the expected file for the integ tests.

File: packages/@aws-cdk/aws-codepipeline-actions/lib/ecs/deploy-action.ts [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+      artifactBounds: {
+        minInputs: 1,
+        maxInputs: 1,
+        minOutputs: 0,
+        maxOutputs: 0,
+      },
+      configuration: {
+        ClusterName: props.service.clusterName,
+        ServiceName: props.service.serviceName,
+        FileName: props.imageFile ? props.imageFile.fileName : undefined,
```
Comment (RomainMuller) [inline packages/@aws-cdk/aws-codepipeline-actions/lib/ecs/deploy-action.ts]: `props.imageFile && props.imageFile.fileName`

File: packages/@aws-cdk/aws-codepipeline-actions/lib/ecs/deploy-action.ts [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+          'ec2.amazonaws.com',
+          'ecs-tasks.amazonaws.com',
+        ],
+      }));
+  }
+}
+
+function determineInputArtifact(props: EcsDeployActionProps): codepipeline.Artifact {
+  if (props.imageFile) {
+    if (props.inputArtifact) {
```
Comment (RomainMuller) [inline packages/@aws-cdk/aws-codepipeline-actions/lib/ecs/deploy-action.ts]: Start with:
```ts
if (!!props.imageFile === !!props.inputArtifact) {
  throw new Error(`Exactly one of impageFile or inputArtifact must be provided, but ${!!props.imageFile ? 'both' : 'none'} was provided.`);
}
```

Then you don't have to nest so deep anymore.

Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline-actions/lib/ecs/deploy-action.ts]: I don't really like that. You have 4 exits from this method, and each has is its own branch, which seems vary natural and readable to me. The nesting level is only 2, which I don't consider a problem.

Your version uses an (in my opinion) non-obvious JS idiom, and forces me to use `!` in the exits that do actually return a value, as the compiler won't be able to tell which of the properties is definitely not `undefined`. I don't think those things are worth doing to save one level of nesting.

File: packages/@aws-cdk/aws-codepipeline-actions/lib/ecs/deploy-action.ts (hunk 10 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline-actions/lib/ecs/deploy-action.ts]: Changed.
